cdap-api-common/src/main/java/io/cdap/cdap/api/data/schema/Schema.java

```diff
@@ -432,17 +435,35 @@ public static Schema unionOf(Schema...schemas) {
    * Creates a {@link Type#UNION UNION} {@link Schema} which represents a union of all the given schemas.
    * The ordering of the schemas inside the union would be the same as the {@link Iterable#iterator()} order.
    *
+   * Unions may not contain more than one schema with the same type, except for records and enums.
+   * For example, unions containing two array types or two map types are not permitted.
+   *
    * @param schemas All the {@link Schema Schemas} constitutes the union.
    * @return A {@link Schema} of {@link Type#UNION UNION} type.
    */
   public static Schema unionOf(Iterable<Schema> schemas) {
+    int numMaps = 0;
+    int numArrays = 0;
     List<Schema> schemaList = new ArrayList<>();
     for (Schema schema : schemas) {
       schemaList.add(schema);
+      Schema.Type type = schema.getType();
+      if (type == Type.MAP) {
+        numMaps++;
+      }
+      if (type == Type.ARRAY) {
+        numArrays++;
+      }
     }
     if (schemaList.isEmpty()) {
       throw new IllegalArgumentException("No union schema provided.");
     }
+    if (numMaps > 1) {
```

Comment:
I would consider this a backward-incompatible bug fix, so it is possible it could break existing code.

Practically speaking I don't think anybody would run into this. Unions are pretty poorly supported in the plugins we package out of the box, and allowing multiple collections opens the door for situations where it is impossible to efficiently determine the value type. For example, if somebody creates a union of a list of non-nullable integers and a list of nullable integers.

Reply:
agreed, I think it is ok.